File: frontend/src/services/backupRestore.ts
import { buildScopeQueryKey, fetchJSON } from '@/services/api/client';

type UnknownRecord = Record<string, unknown>;
type QueryValue = string | number | boolean | undefined | null;

const isRecord = (value: unknown): value is UnknownRecord =>
  typeof value === 'object' && value !== null;

const pick = (record: UnknownRecord, keys: string[]): unknown => {
  for (const key of keys) {
    if (key in record) {
      return record[key];
    }
  }
  return undefined;
};

const toText = (value: unknown): string | undefined => {
  if (typeof value === 'string') {
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : undefined;
  }
  if (typeof value === 'number' && Number.isFinite(value)) {
    return String(value);
  }
  return undefined;
};

const toNumber = (value: unknown): number | undefined => {
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === 'string') {
    const parsed = Number(value.trim());
    return Number.isFinite(parsed) ? parsed : undefined;
  }
  return undefined;
};

const toStringArray = (value: unknown): string[] => {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .map((item) => toText(item))
    .filter((item): item is string => Boolean(item));
};

const toQueryString = (query: Record<string, QueryValue>) => {
  const params = new URLSearchParams();
  Object.entries(query).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') {
      return;
    }
    params.set(key, String(value));
  });
  return params.toString();
};

const withQuery = (path: string, query: Record<string, QueryValue>) => {
  const queryString = toQueryString(query);
  return queryString ? `${path}?${queryString}` : path;
};

const normalizeListResponse = <T>(
  value: unknown,
  mapper: (record: UnknownRecord, index: number) => T
): { items: T[] } => {
  if (Array.isArray(value)) {
    return { items: value.filter(isRecord).map(mapper) };
  }
  const record = isRecord(value) ? value : {};
  const items = Array.isArray(record.items)
    ? record.items
    : Array.isArray(record.Items)
      ? record.Items
      : [];
  return { items: items.filter(isRecord).map(mapper) };
};

export type BackupPolicyStatus = 'active' | 'paused' | 'draft' | 'failed' | string;
export type RestorePointResult = 'succeeded' | 'partial' | 'failed' | 'running' | string;
export type RestoreJobType =
  | 'in-place'
  | 'cross-cluster'
  | 'selective'
  | 'migration'
  | string;
export type RestoreJobStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'partial'
  | 'failed'
  | 'blocked'
  | string;
export type DrillStatus =
  | 'draft'
  | 'scheduled'
  | 'running'
  | 'completed'
  | 'failed'
  | string;

export type ScopeSelection = Record<string, unknown>;

export type BackupPolicy = {
  id: string;
  name: string;
  description?: string;
  scopeType?: string;
  scopeRef?: string;
  executionMode?: string;
  scheduleExpression?: string;
  retentionRule?: string;
  consistencyLevel?: string;
  status?: BackupPolicyStatus;
  ownerUserId?: string;
  createdAt?: string;
  updatedAt?: string;
};

export type CreateBackupPolicyPayload = {
  name: string;
  description?: string;
  scopeType: string;
  scopeRef?: string;
  executionMode: string;
  scheduleExpression?: string;
  retentionRule: string;
  consistencyLevel: string;
};

export type RestorePoint = {
  id: string;
  policyId?: string;
  scopeSnapshot?: ScopeSelection;
  backupStartedAt?: string;
  backupCompletedAt?: string;
  durationSeconds?: number;
  result?: RestorePointResult;
  consistencySummary?: string;
  failureReason?: string;
  storageRef?: string;
  expiresAt?: string;
  createdBy?: string;
};

export type RestoreJob = {
  id: string;
  restorePointId?: string;
  jobType?: RestoreJobType;
  sourceEnvironment?: string;
  targetEnvironment?: string;
  scopeSelection?: ScopeSelection;
  conflictSummary?: string;
  consistencyNotice?: string;
  status?: RestoreJobStatus;
  resultSummary?: string;
  failureReason?: string;
  requestedBy?: string;
  startedAt?: string;
  completedAt?: string;
};

export type CreateRestoreJobPayload = {
  restorePointId: string;
  jobType: RestoreJobType;
  sourceEnvironment?: string;
  targetEnvironment: string;
  scopeSelection: ScopeSelection;
};

export type PrecheckResult = {
  status?: string;
  blockers: string[];
  warnings: string[];
  consistencyNotice?: string;
};

export type MigrationPlan = {
  id: string;
  name: string;
  sourceClusterId?: string;
  targetClusterId?: string;
  scopeSelection?: ScopeSelection;
  mappingRules?: ScopeSelection;
  cutoverSteps: string[];
  status?: string;
  createdBy?: string;
  createdAt?: string;
  updatedAt?: string;
};

export type CreateMigrationPlanPayload = {
  name: string;
  sourceClusterId: string;
  targetClusterId: string;
  scopeSelection: ScopeSelection;
  mappingRules?: ScopeSelection;
  cutoverSteps?: string[];
};

export type DRDrillPlan = {
  id: string;
  name: string;
  description?: string;
  scopeSelection?: ScopeSelection;
  rpoTargetMinutes?: number;
  rtoTargetMinutes?: number;
  roleAssignments: string[];
  cutoverProcedure: string[];
  validationChecklist: string[];
  status?: DrillStatus;
  createdBy?: string;
  createdAt?: string;
  updatedAt?: string;
};

export type CreateDRDrillPlanPayload = {
  name: string;
  description?: string;
  scopeSelection: ScopeSelection;
  rpoTargetMinutes: number;
  rtoTargetMinutes: number;
  roleAssignments?: string[];
  cutoverProcedure: string[];
  validationChecklist: string[];
};

export type DRDrillRecord = {
  id: string;
  planId?: string;
  startedAt?: string;
  completedAt?: string;
  actualRpoMinutes?: number;
  actualRtoMinutes?: number;
  status?: DrillStatus;
  stepResults: string[];
  validationResults: string[];
  incidentNotes?: string;
  executedBy?: string;
};

export type DRDrillReport = {
  id: string;
  drillRecordId?: string;
  goalAssessment?: string;
  gapSummary?: string;
  issuesFound: string[];
  improvementActions: string[];
  publishedAt?: string;
  publishedBy?: string;
};

export type BackupAuditEvent = {
  id: string;
  action?: string;
  actorUserId?: string;
  targetType?: string;
  targetRef?: string;
  scopeSnapshot?: ScopeSelection;
  outcome?: string;
  detailSnapshot?: ScopeSelection;
  occurredAt?: string;
};

export type BackupPolicyListQuery = {
  scopeType?: string;
  status?: string;
};

export type RestorePointListQuery = {
  policyId?: string;
  result?: string;
};

export type RestoreJobListQuery = {
  jobType?: string;
  status?: string;
};

export type BackupAuditListQuery = {
  action?: string;
  outcome?: string;
  targetType?: string;
};

export const backupRestoreQueryKeys = {
  all: ['backup-restore'] as const,
  policies: (scope?: string) => ['backup-restore', 'policies', scope ?? 'all'] as const,
  restorePoints: (scope?: string) =>
    ['backup-restore', 'restore-points', scope ?? 'all'] as const,
  restorePointDetail: (restorePointId?: string) =>
    ['backup-restore', 'restore-point-detail', restorePointId ?? 'unknown'] as const,
  restoreJobs: (scope?: string) =>
    ['backup-restore', 'restore-jobs', scope ?? 'all'] as const,
  drillPlans: (scope?: string) => ['backup-restore', 'drill-plans', scope ?? 'all'] as const,
  drillRecord: (recordId?: string) =>
    ['backup-restore', 'drill-record', recordId ?? 'unknown'] as const,
  audit: (scope?: string) => ['backup-restore', 'audit', scope ?? 'all'] as const
};

const mapBackupPolicy = (record: UnknownRecord): BackupPolicy => ({
  id: toText(pick(record, ['id', 'policyId'])) || 'unknown-policy',
  name: toText(pick(record, ['name', 'displayName'])) || '未命名策略',
  description: toText(record.description),
  scopeType: toText(record.scopeType),
  scopeRef: toText(record.scopeRef),
  executionMode: toText(record.executionMode),
  scheduleExpression: toText(record.scheduleExpression),
  retentionRule: toText(record.retentionRule),
  consistencyLevel: toText(record.consistencyLevel),
  status: toText(record.status),
  ownerUserId: toText(record.ownerUserId),
  createdAt: toText(record.createdAt),
  updatedAt: toText(record.updatedAt)
});

const mapRestorePoint = (record: UnknownRecord): RestorePoint => ({
  id: toText(pick(record, ['id', 'restorePointId'])) || 'unknown-restore-point',
  policyId: toText(record.policyId),
  scopeSnapshot: isRecord(record.scopeSnapshot) ? record.scopeSnapshot : undefined,
  backupStartedAt: toText(record.backupStartedAt),
  backupCompletedAt: toText(record.backupCompletedAt),
  durationSeconds: toNumber(record.durationSeconds),
  result: toText(record.result),
  consistencySummary: toText(record.consistencySummary),
  failureReason: toText(record.failureReason),
  storageRef: toText(record.storageRef),
  expiresAt: toText(record.expiresAt),
  createdBy: toText(record.createdBy)
});

const mapRestoreJob = (record: UnknownRecord): RestoreJob => ({
  id: toText(pick(record, ['id', 'jobId'])) || 'unknown-restore-job',
  restorePointId: toText(record.restorePointId),
  jobType: toText(record.jobType),
  sourceEnvironment: toText(record.sourceEnvironment),
  targetEnvironment: toText(record.targetEnvironment),
  scopeSelection: isRecord(record.scopeSelection) ? record.scopeSelection : undefined,
  conflictSummary: toText(record.conflictSummary),
  consistencyNotice: toText(record.consistencyNotice),
  status: toText(record.status),
  resultSummary: toText(record.resultSummary),
  failureReason: toText(record.failureReason),
  requestedBy: toText(record.requestedBy),
  startedAt: toText(record.startedAt),
  completedAt: toText(record.completedAt)
});

const mapPrecheckResult = (value: unknown): PrecheckResult => {
  const record = isRecord(value) ? value : {};
  return {
    status: toText(record.status),
    blockers: toStringArray(record.blockers),
    warnings: toStringArray(record.warnings),
    consistencyNotice: toText(record.consistencyNotice)
  };
};

const mapMigrationPlan = (record: UnknownRecord): MigrationPlan => ({
  id: toText(pick(record, ['id', 'migrationPlanId'])) || 'unknown-migration-plan',
  name: toText(record.name) || '未命名迁移计划',
  sourceClusterId: toText(record.sourceClusterId),
  targetClusterId: toText(record.targetClusterId),
  scopeSelection: isRecord(record.scopeSelection) ? record.scopeSelection : undefined,
  mappingRules: isRecord(record.mappingRules) ? record.mappingRules : undefined,
  cutoverSteps: toStringArray(record.cutoverSteps),
  status: toText(record.status),
  createdBy: toText(record.createdBy),
  createdAt: toText(record.createdAt),
  updatedAt: toText(record.updatedAt)
});

const mapDrillPlan = (record: UnknownRecord): DRDrillPlan => ({
  id: toText(pick(record, ['id', 'planId'])) || 'unknown-drill-plan',
  name: toText(record.name) || '未命名演练计划',
  description: toText(record.description),
  scopeSelection: isRecord(record.scopeSelection) ? record.scopeSelection : undefined,
  rpoTargetMinutes: toNumber(record.rpoTargetMinutes),
  rtoTargetMinutes: toNumber(record.rtoTargetMinutes),
  roleAssignments: toStringArray(record.roleAssignments),
  cutoverProcedure: toStringArray(record.cutoverProcedure),
  validationChecklist: toStringArray(record.validationChecklist),
  status: toText(record.status),
  createdBy: toText(record.createdBy),
  createdAt: toText(record.createdAt),
  updatedAt: toText(record.updatedAt)
});

const mapDrillRecord = (record: UnknownRecord): DRDrillRecord => ({
  id: toText(pick(record, ['id', 'recordId'])) || 'unknown-drill-record',
  planId: toText(record.planId),
  startedAt: toText(record.startedAt),
  completedAt: toText(record.completedAt),
  actualRpoMinutes: toNumber(record.actualRpoMinutes),
  actualRtoMinutes: toNumber(record.actualRtoMinutes),
  status: toText(record.status),
  stepResults: toStringArray(record.stepResults),
  validationResults: toStringArray(record.validationResults),
  incidentNotes: toText(record.incidentNotes),
  executedBy: toText(record.executedBy)
});

const mapDrillReport = (record: UnknownRecord): DRDrillReport => ({
  id: toText(pick(record, ['id', 'reportId'])) || 'unknown-drill-report',
  drillRecordId: toText(record.drillRecordId),
  goalAssessment: toText(record.goalAssessment),
  gapSummary: toText(record.gapSummary),
  issuesFound: toStringArray(record.issuesFound),
  improvementActions: toStringArray(record.improvementActions),
  publishedAt: toText(record.publishedAt),
  publishedBy: toText(record.publishedBy)
});

const mapAuditEvent = (record: UnknownRecord): BackupAuditEvent => ({
  id: toText(pick(record, ['id', 'eventId'])) || 'unknown-audit-event',
  action: toText(record.action),
  actorUserId: toText(record.actorUserId),
  targetType: toText(record.targetType),
  targetRef: toText(record.targetRef),
  scopeSnapshot: isRecord(record.scopeSnapshot) ? record.scopeSnapshot : undefined,
  outcome: toText(record.outcome),
  detailSnapshot: isRecord(record.detailSnapshot) ? record.detailSnapshot : undefined,
  occurredAt: toText(record.occurredAt)
});

export const listBackupPolicies = async (query: BackupPolicyListQuery) => {
  const response = await fetchJSON<unknown>(withQuery('/backup-restore/policies', query));
  return normalizeListResponse(response, mapBackupPolicy);
};

export const createBackupPolicy = async (payload: CreateBackupPolicyPayload) => {
  const response = await fetchJSON<unknown>('/backup-restore/policies', {
    method: 'POST',
    body: JSON.stringify(payload)
  });
  return mapBackupPolicy(isRecord(response) ? response : {});
};

export const runBackupPolicy = async (policyId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/policies/${policyId}/run`, {
    method: 'POST'
  });
  return mapRestorePoint(isRecord(response) ? response : {});
};

export const listRestorePoints = async (query: RestorePointListQuery) => {
  const response = await fetchJSON<unknown>(withQuery('/backup-restore/restore-points', query));
  return normalizeListResponse(response, mapRestorePoint);
};

export const getRestorePointDetail = async (restorePointId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/restore-points/${restorePointId}`);
  return mapRestorePoint(isRecord(response) ? response : {});
};

export const listRestoreJobs = async (query: RestoreJobListQuery) => {
  const response = await fetchJSON<unknown>(withQuery('/backup-restore/restore-jobs', query));
  return normalizeListResponse(response, mapRestoreJob);
};

export const createRestoreJob = async (payload: CreateRestoreJobPayload) => {
  const response = await fetchJSON<unknown>('/backup-restore/restore-jobs', {
    method: 'POST',
    body: JSON.stringify(payload)
  });
  return mapRestoreJob(isRecord(response) ? response : {});
};

export const validateRestoreJob = async (jobId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/restore-jobs/${jobId}/validate`, {
    method: 'POST'
  });
  return mapPrecheckResult(response);
};

export const createMigrationPlan = async (payload: CreateMigrationPlanPayload) => {
  const response = await fetchJSON<unknown>('/backup-restore/migrations', {
    method: 'POST',
    body: JSON.stringify(payload)
  });
  return mapMigrationPlan(isRecord(response) ? response : {});
};

export const listDrillPlans = async () => {
  const response = await fetchJSON<unknown>('/backup-restore/drills/plans');
  return normalizeListResponse(response, mapDrillPlan);
};

export const createDrillPlan = async (payload: CreateDRDrillPlanPayload) => {
  const response = await fetchJSON<unknown>('/backup-restore/drills/plans', {
    method: 'POST',
    body: JSON.stringify(payload)
  });
  return mapDrillPlan(isRecord(response) ? response : {});
};

export const runDrillPlan = async (planId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/drills/plans/${planId}/run`, {
    method: 'POST'
  });
  return mapDrillRecord(isRecord(response) ? response : {});
};

export const getDrillRecordDetail = async (recordId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/drills/records/${recordId}`);
  return mapDrillRecord(isRecord(response) ? response : {});
};

export const createDrillReport = async (recordId: string) => {
  const response = await fetchJSON<unknown>(`/backup-restore/drills/records/${recordId}/report`, {
    method: 'POST'
  });
  return mapDrillReport(isRecord(response) ? response : {});
};

export const listBackupRestoreAuditEvents = async (query: BackupAuditListQuery) => {
  const response = await fetchJSON<unknown>(withQuery('/audit/backup-restore/events', query));
  return normalizeListResponse(response, mapAuditEvent);
};

export const backupRestoreScopeSummary = (value: ScopeSelection | undefined) => {
  if (!value || Object.keys(value).length === 0) {
    return '未指定范围';
  }
  return Object.entries(value)
    .map(([key, item]) => {
      if (Array.isArray(item)) {
        return `${key}:${item.length}`;
      }
      if (typeof item === 'string') {
        return `${key}:${item}`;
      }
      if (isRecord(item)) {
        return `${key}:${Object.keys(item).length}`;
      }
      return `${key}`;
    })
    .join(' / ');
};

export const restoreJobQueryScope = (query: RestoreJobListQuery) =>
  buildScopeQueryKey([query.jobType, query.status]);

export const policyQueryScope = (query: BackupPolicyListQuery) =>
  buildScopeQueryKey([query.scopeType, query.status]);

export const restorePointQueryScope = (query: RestorePointListQuery) =>
  buildScopeQueryKey([query.policyId, query.result]);

File: backend/internal/domain/backup_restore.go
package domain

import "time"

type BackupPolicyStatus string

const (
	BackupPolicyStatusDraft      BackupPolicyStatus = "draft"
	BackupPolicyStatusActive     BackupPolicyStatus = "active"
	BackupPolicyStatusPaused     BackupPolicyStatus = "paused"
	BackupPolicyStatusDeprecated BackupPolicyStatus = "deprecated"
)

type RestorePointResult string

const (
	RestorePointResultSucceeded          RestorePointResult = "succeeded"
	RestorePointResultPartiallySucceeded RestorePointResult = "partially_succeeded"
	RestorePointResultFailed             RestorePointResult = "failed"
	RestorePointResultExpired            RestorePointResult = "expired"
)

type RestoreJobType string

const (
	RestoreJobTypeInPlace      RestoreJobType = "in-place-restore"
	RestoreJobTypeCrossCluster RestoreJobType = "cross-cluster-restore"
	RestoreJobTypeMigration    RestoreJobType = "environment-migration"
	RestoreJobTypeSelective    RestoreJobType = "selective-restore"
)

type RestoreJobStatus string

const (
	RestoreJobStatusPending            RestoreJobStatus = "pending"
	RestoreJobStatusValidating         RestoreJobStatus = "validating"
	RestoreJobStatusRunning            RestoreJobStatus = "running"
	RestoreJobStatusSucceeded          RestoreJobStatus = "succeeded"
	RestoreJobStatusPartiallySucceeded RestoreJobStatus = "partially_succeeded"
	RestoreJobStatusFailed             RestoreJobStatus = "failed"
	RestoreJobStatusCanceled           RestoreJobStatus = "canceled"
	RestoreJobStatusBlocked            RestoreJobStatus = "blocked"
)

type MigrationPlanStatus string

const (
	MigrationPlanStatusDraft     MigrationPlanStatus = "draft"
	MigrationPlanStatusApproved  MigrationPlanStatus = "approved"
	MigrationPlanStatusRunning   MigrationPlanStatus = "running"
	MigrationPlanStatusSucceeded MigrationPlanStatus = "succeeded"
	MigrationPlanStatusFailed    MigrationPlanStatus = "failed"
	MigrationPlanStatusCanceled  MigrationPlanStatus = "canceled"
)

type DRDrillPlanStatus string

const (
	DRDrillPlanStatusDraft   DRDrillPlanStatus = "draft"
	DRDrillPlanStatusActive  DRDrillPlanStatus = "active"
	DRDrillPlanStatusPaused  DRDrillPlanStatus = "paused"
	DRDrillPlanStatusRetired DRDrillPlanStatus = "retired"
)

type DRDrillRecordStatus string

const (
	DRDrillRecordStatusPending            DRDrillRecordStatus = "pending"
	DRDrillRecordStatusRunning            DRDrillRecordStatus = "running"
	DRDrillRecordStatusSucceeded          DRDrillRecordStatus = "succeeded"
	DRDrillRecordStatusFailed             DRDrillRecordStatus = "failed"
	DRDrillRecordStatusPartiallySucceeded DRDrillRecordStatus = "partially_succeeded"
	DRDrillRecordStatusCanceled           DRDrillRecordStatus = "canceled"
)

type BackupAuditOutcome string

const (
	BackupAuditOutcomeSucceeded BackupAuditOutcome = "succeeded"
	BackupAuditOutcomeFailed    BackupAuditOutcome = "failed"
	BackupAuditOutcomeBlocked   BackupAuditOutcome = "blocked"
	BackupAuditOutcomeCanceled  BackupAuditOutcome = "canceled"
)

type BackupPolicy struct {
	ID                 uint64             `gorm:"primaryKey" json:"id"`
	Name               string             `gorm:"size:128;not null;index:idx_backup_policy_scope_name,priority:3" json:"name"`
	Description        string             `gorm:"type:text" json:"description,omitempty"`
	ScopeType          string             `gorm:"size:64;not null;index:idx_backup_policy_scope_name,priority:1" json:"scopeType"`
	ScopeRef           string             `gorm:"size:128;not null;index:idx_backup_policy_scope_name,priority:2" json:"scopeRef"`
	WorkspaceID        uint64             `gorm:"not null;index" json:"workspaceId"`
	ProjectID          *uint64            `gorm:"index" json:"projectId,omitempty"`
	ExecutionMode      string             `gorm:"size:32;not null" json:"executionMode"`
	ScheduleExpression string             `gorm:"size:256" json:"scheduleExpression,omitempty"`
	RetentionRule      string             `gorm:"size:256;not null" json:"retentionRule"`
	ConsistencyLevel   string             `gorm:"size:64;not null" json:"consistencyLevel"`
	Status             BackupPolicyStatus `gorm:"size:32;not null;index" json:"status"`
	OwnerUserID        uint64             `gorm:"not null;index" json:"ownerUserId"`
	CreatedAt          time.Time          `json:"createdAt"`
	UpdatedAt          time.Time          `json:"updatedAt"`
}

func (BackupPolicy) TableName() string { return "backup_policies" }

type RestorePoint struct {
	ID                 uint64             `gorm:"primaryKey" json:"id"`
	PolicyID           uint64             `gorm:"not null;index" json:"policyId"`
	WorkspaceID        uint64             `gorm:"not null;index" json:"workspaceId"`
	ProjectID          *uint64            `gorm:"index" json:"projectId,omitempty"`
	ScopeSnapshot      string             `gorm:"type:text;not null" json:"scopeSnapshot"`
	BackupStartedAt    time.Time          `gorm:"not null" json:"backupStartedAt"`
	BackupCompletedAt  *time.Time         `json:"backupCompletedAt,omitempty"`
	DurationSeconds    int                `gorm:"not null;default:0" json:"durationSeconds"`
	Result             RestorePointResult `gorm:"size:32;not null;index" json:"result"`
	ConsistencySummary string             `gorm:"type:text" json:"consistencySummary,omitempty"`
	FailureReason      string             `gorm:"type:text" json:"failureReason,omitempty"`
	StorageRef         string             `gorm:"size:256" json:"storageRef,omitempty"`
	ExpiresAt          *time.Time         `json:"expiresAt,omitempty"`
	CreatedBy          uint64             `gorm:"not null;index" json:"createdBy"`
	CreatedAt          time.Time          `json:"createdAt"`
	UpdatedAt          time.Time          `json:"updatedAt"`
}

func (RestorePoint) TableName() string { return "restore_points" }

type RestoreJob struct {
	ID                uint64           `gorm:"primaryKey" json:"id"`
	RestorePointID    uint64           `gorm:"not null;index" json:"restorePointId"`
	WorkspaceID       uint64           `gorm:"not null;index" json:"workspaceId"`
	ProjectID         *uint64          `gorm:"index" json:"projectId,omitempty"`
	JobType           RestoreJobType   `gorm:"size:64;not null;index" json:"jobType"`
	SourceEnvironment string           `gorm:"size:128" json:"sourceEnvironment,omitempty"`
	TargetEnvironment string           `gorm:"size:128;not null" json:"targetEnvironment"`
	ScopeSelection    string           `gorm:"type:text;not null" json:"scopeSelection"`
	ConflictSummary   string           `gorm:"type:text" json:"conflictSummary,omitempty"`
	ConsistencyNotice string           `gorm:"type:text" json:"consistencyNotice,omitempty"`
	Status            RestoreJobStatus `gorm:"size:32;not null;index" json:"status"`
	ResultSummary     string           `gorm:"type:text" json:"resultSummary,omitempty"`
	FailureReason     string           `gorm:"type:text" json:"failureReason,omitempty"`
	RequestedBy       uint64           `gorm:"not null;index" json:"requestedBy"`
	StartedAt         *time.Time       `json:"startedAt,omitempty"`
	CompletedAt       *time.Time       `json:"completedAt,omitempty"`
	CreatedAt         time.Time        `json:"createdAt"`
	UpdatedAt         time.Time        `json:"updatedAt"`
}

func (RestoreJob) TableName() string { return "restore_jobs" }

type MigrationPlan struct {
	ID              uint64              `gorm:"primaryKey" json:"id"`
	Name            string              `gorm:"size:128;not null;uniqueIndex" json:"name"`
	WorkspaceID     uint64              `gorm:"not null;index" json:"workspaceId"`
	ProjectID       *uint64             `gorm:"index" json:"projectId,omitempty"`
	SourceClusterID uint64              `gorm:"not null;index" json:"sourceClusterId"`
	TargetClusterID uint64              `gorm:"not null;index" json:"targetClusterId"`
	ScopeSelection  string              `gorm:"type:text;not null" json:"scopeSelection"`
	MappingRules    string              `gorm:"type:text" json:"mappingRules,omitempty"`
	CutoverSteps    string              `gorm:"type:text" json:"cutoverSteps,omitempty"`
	Status          MigrationPlanStatus `gorm:"size:32;not null;index" json:"status"`
	CreatedBy       uint64              `gorm:"not null;index" json:"createdBy"`
	CreatedAt       time.Time           `json:"createdAt"`
	UpdatedAt       time.Time           `json:"updatedAt"`
}

func (MigrationPlan) TableName() string { return "migration_plans" }

type DRDrillPlan struct {
	ID                  uint64            `gorm:"primaryKey" json:"id"`
	Name                string            `gorm:"size:128;not null;uniqueIndex" json:"name"`
	Description         string            `gorm:"type:text" json:"description,omitempty"`
	WorkspaceID         uint64            `gorm:"not null;index" json:"workspaceId"`
	ProjectID           *uint64           `gorm:"index" json:"projectId,omitempty"`
	ScopeSelection      string            `gorm:"type:text;not null" json:"scopeSelection"`
	RPOTargetMinutes    int               `gorm:"not null" json:"rpoTargetMinutes"`
	RTOTargetMinutes    int               `gorm:"not null" json:"rtoTargetMinutes"`
	RoleAssignments     string            `gorm:"type:text" json:"roleAssignments,omitempty"`
	CutoverProcedure    string            `gorm:"type:text;not null" json:"cutoverProcedure"`
	ValidationChecklist string            `gorm:"type:text;not null" json:"validationChecklist"`
	Status              DRDrillPlanStatus `gorm:"size:32;not null;index" json:"status"`
	CreatedBy           uint64            `gorm:"not null;index" json:"createdBy"`
	CreatedAt           time.Time         `json:"createdAt"`
	UpdatedAt           time.Time         `json:"updatedAt"`
}

func (DRDrillPlan) TableName() string { return "dr_drill_plans" }

type DRDrillRecord struct {
	ID                uint64              `gorm:"primaryKey" json:"id"`
	PlanID            uint64              `gorm:"not null;index" json:"planId"`
	WorkspaceID       uint64              `gorm:"not null;index" json:"workspaceId"`
	ProjectID         *uint64             `gorm:"index" json:"projectId,omitempty"`
	StartedAt         time.Time           `gorm:"not null" json:"startedAt"`
	CompletedAt       *time.Time          `json:"completedAt,omitempty"`
	ActualRPOMinutes  int                 `gorm:"not null;default:0" json:"actualRpoMinutes"`
	ActualRTOMinutes  int                 `gorm:"not null;default:0" json:"actualRtoMinutes"`
	Status            DRDrillRecordStatus `gorm:"size:32;not null;index" json:"status"`
	StepResults       string              `gorm:"type:text" json:"stepResults,omitempty"`
	ValidationResults string              `gorm:"type:text" json:"validationResults,omitempty"`
	IncidentNotes     string              `gorm:"type:text" json:"incidentNotes,omitempty"`
	ExecutedBy        uint64              `gorm:"not null;index" json:"executedBy"`
	CreatedAt         time.Time           `json:"createdAt"`
	UpdatedAt         time.Time           `json:"updatedAt"`
}

func (DRDrillRecord) TableName() string { return "dr_drill_records" }

type DRDrillReport struct {
	ID                 uint64    `gorm:"primaryKey" json:"id"`
	DrillRecordID      uint64    `gorm:"not null;uniqueIndex" json:"drillRecordId"`
	GoalAssessment     string    `gorm:"type:text;not null" json:"goalAssessment"`
	GapSummary         string    `gorm:"type:text" json:"gapSummary,omitempty"`
	IssuesFound        string    `gorm:"type:text" json:"issuesFound,omitempty"`
	ImprovementActions string    `gorm:"type:text;not null" json:"improvementActions"`
	PublishedAt        time.Time `gorm:"not null" json:"publishedAt"`
	PublishedBy        uint64    `gorm:"not null;index" json:"publishedBy"`
	CreatedAt          time.Time `json:"createdAt"`
	UpdatedAt          time.Time `json:"updatedAt"`
}

func (DRDrillReport) TableName() string { return "dr_drill_reports" }

type BackupAuditEvent struct {
	ID             uint64             `gorm:"primaryKey" json:"id"`
	Action         string             `gorm:"size:128;not null;index" json:"action"`
	ActorUserID    uint64             `gorm:"not null;index" json:"actorUserId"`
	TargetType     string             `gorm:"size:64;not null;index" json:"targetType"`
	TargetRef      string             `gorm:"size:128;not null;index" json:"targetRef"`
	WorkspaceID    uint64             `gorm:"not null;index" json:"workspaceId"`
	ProjectID      *uint64            `gorm:"index" json:"projectId,omitempty"`
	ScopeSnapshot  string             `gorm:"type:text" json:"scopeSnapshot,omitempty"`
	Outcome        BackupAuditOutcome `gorm:"size:32;not null;index" json:"outcome"`
	DetailSnapshot string             `gorm:"type:text" json:"detailSnapshot,omitempty"`
	OccurredAt     time.Time          `gorm:"not null;index" json:"occurredAt"`
	CreatedAt      time.Time          `json:"createdAt"`
	UpdatedAt      time.Time          `json:"updatedAt"`
}

func (BackupAuditEvent) TableName() string { return "backup_audit_events" }
